### PR TITLE
Ticket reset

### LINKED ISSUE #188: Resetting tickets (and their children)
For our use case, it's useful to have resetTicket methode in the API.
To allow re-encode a video without re-creating the encoding ticket.
The reset should set a ticket back to its initial state.
So after that commenceState has to be called to get things happen again.
When resetting a meta ticket all encoding childs should also be reset.
When resetting a 'master' encoding ticket, all tickets of a depending profile should be reset too.

Do we implement this as a model function or stored procedure?

### PR DESCRIPTION
**Pull request by j-licht**
Add SQL and XMLRPC function to reset a ticket.
If the ticket is a meta ticket all child will reset.
If the ticket is a encoding ticket all encoding tickets will reset depending on the given.

Fixes #188

### CODE REVIEW

**Comment by jjeising** [comment]
This seems to be the first time we log inside stored procedures. Would it be possible to move these to the Model?

**Comment by j-licht** [comment]
> This seems to be the first time we log inside stored procedures. Would it be possible to move these to the Model?

One possibility would be change the new function to collect and return ticket id, from and to state.
And in the model we can iterate over that and create the log entries.

File: src/Application/Migrations/23_function_encoding_profile_all_dependee.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    param_known_ids := array[param_profile_id];
+  END IF;
+
+  FOR p IN
+    SELECT p.id
+    FROM tbl_encoding_profile p
+    WHERE depends_on=param_profile_id AND NOT (p.id = ANY(param_known_ids))
+  LOOP
+    ids := p || ids;
+    ids := (SELECT ARRAY(SELECT e.id from encoding_profile_all_dependees(p, ids || param_known_ids) e)) || ids;
```
Comment (pegro) [inline src/Application/Migrations/23_function_encoding_profile_all_dependee.sql]: Mh, I thought this would creates duplicates when returning from recursion, but with an example I tried on paper I coulnd't reproduce this, so I guess this is just correct.

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        JOIN tbl_encoding_profile_version epv ON tt.encoding_profile_version_id = epv.id
+        WHERE tt.id = param_ticket_id
+      ) pd ON pd.id = epv2.encoding_profile_id AND t2.parent_id = pd.parent_id
+    LOOP
+      inital_state := ticket_state_initial(t.project_id, t.ticket_type);
+      IF inital_state <> t.ticket_state THEN
+        INSERT INTO tbl_log (
+          ticket_id, from_state, to_state, handle_id, event, comment
+          ) VALUES (t.id, t.ticket_state, inital_state, param_handle_id, 'Action.reset', 'Following encoding ticket.');
+        UPDATE tbl_ticket SET ticket_state=inital_state WHERE id=t.id;
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: Try to be consistent with spacing around operators. I personally like spaces before and after them.

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 3]
```diff
@@ -0,0 +1,70 @@
+BEGIN;
+
+--param_handle_id is used for the log entry
+CREATE OR REPLACE FUNCTION ticket_reset(param_ticket_id bigint, param_handle_id bigint)
+  RETURNS void AS
+$$
+DECLARE
+  ticket tbl_ticket%rowtype;
+  t tbl_ticket%rowtype;
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: It looks like you could use the "ticket" variable in all cases, so you shouldn't need another variable.

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  inital_state enum_ticket_state;
+BEGIN
+  --first get the full given ticket
+  SELECT * INTO ticket FROM tbl_ticket where id = param_ticket_id;
+
+  IF ticket IS NULL THEN
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: As stated in https://www.postgresql.org/docs/current/static/plpgsql-statements.html#PLPGSQL-STATEMENTS-DIAGNOSTICS you can check, whether the variable FOUND is set to false.

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  inital_state := ticket_state_initial(ticket.project_id, ticket.ticket_type);
+  IF inital_state <> ticket.ticket_state THEN
+    INSERT INTO tbl_log (
+      ticket_id, from_state, to_state, handle_id, event
+      ) VALUES (ticket.id, ticket.ticket_state, inital_state, param_handle_id, 'Action.reset');
+    UPDATE tbl_ticket SET ticket_state=inital_state WHERE id=param_ticket_id;
+  END IF;
+
+  --if given ticket is a meta ticket, reset all childs
+  IF ticket.ticket_type = 'meta'::enum_ticket_type THEN
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: I think you can remove the type check (which I don't really like in SQL prodecures anyway), and rather reset all tickets WHERE parent = ticket.id unconditionally. Other types should just have no children, so there is no harm doing that.

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      IF inital_state <> t.ticket_state THEN
+        INSERT INTO tbl_log (
+          ticket_id, from_state, to_state, handle_id, event, comment
+          ) VALUES (t.id, t.ticket_state, inital_state, param_handle_id, 'Action.reset', 'Following meta Ticket.');
+        UPDATE tbl_ticket SET ticket_state=inital_state WHERE id=t.id;
+      END IF;
+    END LOOP;
+
+  --if given ticket is an encoding ticket , reset all ticket depends on it by encoding profile
+  ELSEIF ticket.ticket_type = 'encoding'::enum_ticket_type THEN
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: An here I'd rather just check, if ticket.encoding_profile_version_id IS NOT NULL, since that's the important column in the join clause, but technically you could even do this query conditionally too. For other ticket types this loop should do nothing.

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    FOR t IN
+    SELECT t2.* FROM tbl_ticket t2
+      JOIN tbl_encoding_profile_version epv2 ON t2.encoding_profile_version_id = epv2.id
+      JOIN (
+        SELECT
+          encoding_profile_all_dependees(epv.encoding_profile_id) AS id,
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: Have you actually tested this code?
I mean I could be wrong, but the function "encoding_profile_all_dependees" returns a query/table, not a scalar type. So I would expect it in the FROM clause, not as a column in the SELECT clause.

File: src/Application/Model/LogEntry.php [hunk 8]
```diff
@@ -126,6 +126,12 @@ class LogEntry extends Model {
 				'multiple' => '{user_name} abortet checking {tickets}.',
 				'message' => false
 			],
+			'Action.reset' => [
+				'log' => 'Ticket reset.',
+				'single' => '{user_name} rested {id} to state {to_state}',
+				'multiple' => '{user_name} rested {tickets}',
```
Comment (pegro) [inline src/Application/Model/LogEntry.php]: {user_name} reset {id}/{tickets}
Past tense of reset is reset

File: src/Application/Model/LogEntry.php [hunk 9]
```diff
@@ -126,6 +126,12 @@ class LogEntry extends Model {
 				'multiple' => '{user_name} abortet checking {tickets}.',
```
Comment (pegro) [inline src/Application/Model/LogEntry.php]: Unrelated, but "abortet" might also not be an English word ;)

File: src/Application/Migrations/23_function_encoding_profile_all_dependee.sql [hunk 10]
```diff
@@ -0,0 +1,26 @@
+BEGIN;
+
+CREATE OR REPLACE FUNCTION encoding_profile_all_dependees(param_profile_id bigint, param_known_ids bigint[] DEFAULT array[]::bigint[])
+  RETURNS TABLE (id bigint) AS
+$$
+DECLARE
+  p bigint;
```
Comment (pegro) [inline src/Application/Migrations/23_function_encoding_profile_all_dependee.sql]: "p" doesn't have an obvious meaning.
at least "p_id" would be nice or "dependee_profile_id" or similar

File: src/Application/Migrations/23_function_encoding_profile_all_dependee.sql [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  ids bigint[];
```
Comment (pegro) [inline src/Application/Migrations/23_function_encoding_profile_all_dependee.sql]: dependee_ids?
dependee_profile_ids?
profile_ids?

File: src/Application/Migrations/23_function_encoding_profile_all_dependee.sql (hunk 1 above)
Comment (pegro) [reply src/Application/Migrations/23_function_encoding_profile_all_dependee.sql]: Does it have to have that outer SELECT? Does ids := ARRAY() || ids throw errors?

File: src/Application/Migrations/23_function_encoding_profile_all_dependee.sql [hunk 12]
```diff
@@ -0,0 +1,26 @@
+BEGIN;
+
+CREATE OR REPLACE FUNCTION encoding_profile_all_dependees(param_profile_id bigint, param_known_ids bigint[] DEFAULT array[]::bigint[])
```
Comment (pegro) [inline src/Application/Migrations/23_function_encoding_profile_all_dependee.sql]: Needs a description what it does.
Also it should be noted, that the second parameter is optional (and will be left empty in most use cases).

File: src/Application/Model/LogEntry.php (hunk 9 above)
Comment (jjeising) [reply src/Application/Model/LogEntry.php]: Fixed in 39020caf.

File: src/Application/Migrations/24_function_ticket_reset.sql (hunk 4 above)
Comment (pegro) [reply src/Application/Migrations/24_function_ticket_reset.sql]: See paragraph starting with "The second method to determine the effects of a command..."

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  --reset the given ticket to inital state
+  inital_state := ticket_state_initial(ticket.project_id, ticket.ticket_type);
+  IF inital_state <> ticket.ticket_state THEN
+    INSERT INTO tbl_log (
+      ticket_id, from_state, to_state, handle_id, event
+      ) VALUES (ticket.id, ticket.ticket_state, inital_state, param_handle_id, 'Action.reset');
+    UPDATE tbl_ticket SET ticket_state=inital_state WHERE id = param_ticket_id;
+  END IF;
+
+  FOR ticket IN
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: The comment describing what is going on here was nice, it didn't have to vanish, only get updated. (same below)

Comment (pegro) [reply src/Application/Migrations/24_function_ticket_reset.sql]: The plural of child is children ;)

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+      from_state := ticket.ticket_state;
+      to_state := inital_state;
+      follow_meta_ticket := TRUE;
+      follow_encoding_ticket := FALSE;
+      RETURN NEXT;
+      UPDATE tbl_ticket SET (ticket_state, failed, handle_id) = (inital_state, FALSE, NULL) WHERE id = ticket.id;
+    END IF;
+  END LOOP;
+
+  --reset all ticket depends on given ticket by encoding profile
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: reset all tickets with encoding profiles depending on given ticket's profile

File: src/Application/Migrations/24_function_ticket_reset.sql (hunk 7 above)
Comment (pegro) [reply src/Application/Migrations/24_function_ticket_reset.sql]: Ping?

File: src/Application/Model/Ticket.php [hunk 15]
```diff
@@ -851,6 +851,33 @@ public function resetSource(Comment $comment = null) {
...
+				$entry = [
+					'ticket_id' => $ticket['ticket_id'],
+					'from_state' => $ticket['from_state'],
+					'to_state' => $tickets['to_state'],
+					'handle_id' => ($handle_id === null) ? User::getCurrent()['id'] : $handle_id,
+					'event' => 'Action.reset'
+					];
+
+				if ($ticket['follow_meta_ticket']) {
+					$entry['comment'] = 'Following meta ticket.';
```
Comment (pegro) [inline src/Application/Model/Ticket.php]: Mh, not sure about the phrasing here. Maybe "Reset due to meta ticket reset" or something similar is better? @jjeising ?

File: src/Application/Model/Ticket.php [hunk 16]
```diff
@@ -851,6 +851,33 @@ public function resetSource(Comment $comment = null) {
...
+				}
+				if ($ticket['follow_encoding_ticket']) {
+					$entry['comment'] = 'Following encoding ticket.';
```
Comment (pegro) [inline src/Application/Model/Ticket.php]: "Reset due to ticket reset of depending encoding profile"?

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+  --reset the given ticket to inital state
+  inital_state := ticket_state_initial(ticket.project_id, ticket.ticket_type);
+  IF inital_state <> ticket.ticket_state THEN
+    ticket_id := ticket.id;
+    from_state := ticket.ticket_state;
+    to_state := inital_state;
+    follow_meta_ticket := FALSE;
+    follow_encoding_ticket := FALSE;
+    RETURN NEXT;
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: I guess the UPDATE query should happen before the next row is returned, otherwise log entries might be created noting state changes which might not have happened due to errors?!
(same 2 times below)

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 18]
```diff
@@ -0,0 +1,76 @@
+BEGIN;
+
+--param_handle_id is used for the log entry
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: I guess this comment is obsolete now.

File: src/Application/Migrations/24_function_ticket_reset.sql (hunk 7 above)
Comment (pegro) [reply src/Application/Migrations/24_function_ticket_reset.sql]: Mh, I tried for myself and it works. Apparently, you can use function returning sets of rows within the query list.
See https://www.postgresql.org/docs/11/xfunc-sql.html#XFUNC-SQL-FUNCTIONS-RETURNING-SET
"Functions returning sets can also be called in the select list of a query. For each row that the query generates by itself, the set-returning function is invoked, and an output row is generated for each element of the function's result set"

Didn't know that.

File: src/Application/Controller/XMLRPC/Handler.php [hunk 19]
```diff
@@ -254,6 +254,28 @@ public function getTicketInfo($ticket_id)
...
+		 */
+		public function resetTicket($ticket_id)
+		{
+			$ticket = Ticket::find(['id' => $ticket_id]);
+			if (!$ticket) {
+				throw new Exception(__FUNCTION__ . ': ticket not found', 201);
+			}
+
+			if (!in_array($ticket['project_id'], $this->_assignedProjects)) {
+				throw new Exception(__FUNCTION__ . ': ticket in project not assigned to worker group', 202);
```
Comment (pegro) [inline src/Application/Controller/XMLRPC/Handler.php]: The error code should be unique within this controller.

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 20]
```diff
@@ -0,0 +1,75 @@
+BEGIN;
+
+CREATE OR REPLACE FUNCTION ticket_reset(param_ticket_id bigint)
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: Mh, for a future extension of this function, it would be nice to limit the types of tickets to be reset.
When you have a recording ticket, you might not want to reset that to "scheduled" once it moved beyond the "recorded" state.
Maybe as a optional second parameter "enum_ticket_type[] param_ticket_types"

File: src/Application/Controller/XMLRPC/Handler.php [hunk 21]
```diff
@@ -254,6 +254,28 @@ public function getTicketInfo($ticket_id)
...
+			}
+
+			$ticket->reset($this->worker['id']);
```
Comment (pegro) [inline src/Application/Controller/XMLRPC/Handler.php]: The function should return something, at least a boolean.
I know that "no tickets were reset since all matching tickets are already in their initial state" is a valid result of this method. So maybe just return "true" for the time being until we need something more complex.

File: src/Application/Controller/XMLRPC/Handler.php [hunk 22]
```diff
@@ -254,6 +254,28 @@ public function getTicketInfo($ticket_id)
 			
 			return $this->_getTicketInfo($ticket);
 		}
+
+		/**
+		 * Reset given ticket by ticket id to initial state.
+		 * Also reset the childs of a meta ticket and
```
Comment (pegro) [inline src/Application/Controller/XMLRPC/Handler.php]: _children_

The "by ticket id" is a bit redundant with "given ticket"

File: src/Application/Model/Ticket.php [hunk 23]
```diff
@@ -851,6 +851,33 @@ public function resetSource(Comment $comment = null) {
...
+
+				if ($ticket['follow_meta_ticket']) {
+					$entry['comment'] = 'Reset due to meta ticket reset.';
+				}
+				if ($ticket['follow_encoding_ticket']) {
+					$entry['comment'] = 'Reset due to ticket reset of depending encoding profile';
+				}
+
+				LogEntry::create($entry);
+			}
```
Comment (pegro) [inline src/Application/Model/Ticket.php]: I not sure, what this method should return. Maybe `count($tickets)` so when used in the UI, it could present a message like "[n] tickets were reset"?! @jjeising

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 24]
```diff
@@ -0,0 +1,75 @@
+BEGIN;
+
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: Also, it should be documented, that this function resets handle_id and failed flag additionally to resetting the state.

File: src/Application/Migrations/24_function_ticket_reset.sql [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+  IF NOT FOUND THEN
+    RAISE WARNING 'ticket with id % does not exist', param_ticket_id;
+    RETURN;
+  END IF;
+
+  --reset the given ticket to inital state
+  inital_state := ticket_state_initial(ticket.project_id, ticket.ticket_type);
+  IF inital_state <> ticket.ticket_state THEN
+    UPDATE tbl_ticket SET ticket_state=inital_state WHERE id = param_ticket_id;
```
Comment (pegro) [inline src/Application/Migrations/24_function_ticket_reset.sql]: I think handle_id and failed flag should be always reset, not only if the ticket was a child/dependee of someone.

**Comment by pegro** [comment]
Basic functionality works, just small issues left.

**Comment by jjeising** [comment]
Is this pull request ready for a last review?

File: src/Application/Migrations/15_function_ticket_state.sql [hunk 26]
```diff
@@ -129,4 +129,109 @@ END
...
+  SELECT t2.* FROM tbl_ticket t2
+                     JOIN tbl_encoding_profile_version epv2 ON t2.encoding_profile_version_id = epv2.id
+                     JOIN (
+                          SELECT
+                                 encoding_profile_all_dependees(epv.encoding_profile_id) AS id,
+                                 tt.parent_id AS parent_id
+                          FROM tbl_ticket tt
+                                 JOIN tbl_encoding_profile_version epv ON tt.encoding_profile_version_id = epv.id
+                          WHERE tt.id = param_ticket_id
+                          ) pd ON pd.id = epv2.encoding_profile_id AND t2.parent_id = pd.parent_id
```
Comment (pegro) [inline src/Application/Migrations/15_function_ticket_state.sql]: nit: I think the indentation of all JOIN clauses is not necessary, it should be on the same level as FROM and WHERE.

File: src/Application/Migrations/15_function_ticket_state.sql [hunk 27]
```diff
@@ -129,4 +129,109 @@ END
 $$
 LANGUAGE plpgsql;
 
+-- function returns a table with all encoding profiles, which depend
+-- on the given encoding profile
+
+-- second parameter is optional and is used in recursive to prevent endless loop
+
+CREATE OR REPLACE FUNCTION encoding_profile_all_dependees(param_profile_id bigint, param_known_ids bigint[] DEFAULT array[]::bigint[])
```
Comment (pegro) [inline src/Application/Migrations/15_function_ticket_state.sql]: This function doesn't really belong in this file. It's not about ticket states.

File: src/Application/Migrations/23_function_encoding_profile_all_dependee.sql [hunk 28]
```diff
@@ -0,0 +1,31 @@
+BEGIN;
+
```
[the remark this answers is not in the document]
Comment (pegro) [reply src/Application/Migrations/23_function_encoding_profile_all_dependee.sql]: Oh sorry, I meant the function_ticket_dependency_* files.
And the new function kind of fits there as well.

File: src/Application/Migrations/24_function_ticket_reset.sql (hunk 20 above)
Comment (pegro) [reply src/Application/Migrations/24_function_ticket_reset.sql]: Ok, there parameter is there now, but neither used in the model nor in the stored procedure either.
That's kind of pointless. Maybe add it as `-- TODO: ` comment for later, or implement the filter.

**Comment by pegro** [comment]
@jjeising  Yeah, it looks good now. Your review would be welcome.
3 of my comments are left, but that's mostly cosmetic stuff.

File: src/Application/Model/Ticket.php [hunk 29]
```diff
@@ -851,6 +851,34 @@ public function resetSource(Comment $comment = null) {
...
 			return true;
 		}
+
+		public function reset($handle_id = null) {
+			$query = Database::$Instance->query(
+				'SELECT * FROM ticket_reset(?)',
+				[$this['id'], ]
+			);
+
+			$tickets = $query->fetchAll();
```
Comment (jjeising) [inline src/Application/Model/Ticket.php]: Nit: Is `$query` needed here?

File: src/Application/Migrations/15_function_ticket_state.sql [hunk 30]
```diff
@@ -129,4 +129,109 @@ END
...
+    dependee_profile_ids := ARRAY(SELECT e.id from encoding_profile_all_dependees(p_id, dependee_profile_ids || param_known_ids) e) || dependee_profile_ids;
+  END LOOP;
+  RETURN QUERY SELECT DISTINCT unnest(dependee_profile_ids);
+END
+$$
+LANGUAGE plpgsql;
+
+-- function also resets handle_id and failed flag
+CREATE OR REPLACE FUNCTION ticket_reset(param_ticket_id bigint, param_ticket_types enum_ticket_type[] DEFAULT array[]::enum_ticket_type[])
+  RETURNS TABLE (ticket_id bigint,  from_state enum_ticket_state, to_state enum_ticket_state, follow_meta_ticket boolean, follow_encoding_ticket boolean) AS
```
Comment (jjeising) [inline src/Application/Migrations/15_function_ticket_state.sql]: I would probably use _follows_ or _followed_ instead of _follow_, as the later denotes a future action or instruction, but this returns something already happened.

File: src/Application/Model/Ticket.php (hunk 23 above)
Comment (jjeising) [reply src/Application/Model/Ticket.php]: Looks reasonable for the moment.

File: src/Application/Migrations/15_function_ticket_state.sql [hunk 31]
```diff
@@ -129,4 +129,88 @@ END
...
+    SELECT
+      t2.*
+    FROM
+      tbl_ticket t2
+    JOIN
+      tbl_encoding_profile_version epv2 ON t2.encoding_profile_version_id = epv2.id
+    JOIN (
+      SELECT
+        encoding_profile_all_dependees(epv.encoding_profile_id) AS id,
+        tt.parent_id AS parent_id
```
Comment (pegro) [inline src/Application/Migrations/15_function_ticket_state.sql]: If the column has a same name, there is no need for an alias.
Also, I think you don't need the table alias tt here at all, since it's not used outside the subquery and the tables is also not joined with itself.

File: src/Application/Migrations/16_function_ticket_dependency.sql [hunk 32]
```diff
@@ -29,4 +29,33 @@ END
 $$
 LANGUAGE plpgsql;
 
+-- function returns a table with all encoding profiles, which depend
```
Comment (pegro) [inline src/Application/Migrations/16_function_ticket_dependency.sql]: Mh, I changed my mind. "Ticket" is not part of this query in any way, but the file name implies that. So it doesn't fit here either. I guess it's time for a 19_function_encodingprofile_dependency.sql or something like that.

Not part of this PR, but I'd propose to merge all ticket dependency related functions into a single file and this migration then to be renamed.

**Comment by j-licht** [comment]
next try ;)

File: src/Application/Controller/XMLRPC/Handler.php [hunk 33]
```diff
@@ -254,6 +254,30 @@ public function getTicketInfo($ticket_id)
...
+		/**
+		 * Reset given ticket to initial state.
+		 * Also reset the children of a meta ticket and
+		 * encoding tickets which depend on the given ticket.
+		 *
+		 * @param integer $ticket_id ticket identifier
+		 * @throws Exception
+		 * @return boolean
+		 */
+		public function resetTicket($ticket_id)
```
Comment (pegro) [inline src/Application/Controller/XMLRPC/Handler.php]: Last remark, or more of a question: Do you think, now that you implemented this nice ticket type filter, it would be also to expose it via the API?
Like a resetTicket($ticket_id, array $ticket_types = []) method?

File: src/Application/Migrations/15_function_ticket_state.sql [hunk 34]
```diff
@@ -129,4 +129,88 @@ END
...
+-- function also resets handle_id and failed flag
+CREATE OR REPLACE FUNCTION ticket_reset(param_ticket_id bigint, param_ticket_types enum_ticket_type[] DEFAULT array[]::enum_ticket_type[])
+  RETURNS TABLE (ticket_id bigint,  from_state enum_ticket_state, to_state enum_ticket_state, follows_meta_ticket boolean, follows_encoding_ticket boolean) AS
+$$
+DECLARE
+  ticket tbl_ticket%rowtype;
+  inital_state enum_ticket_state;
+BEGIN
+  --first get the full given ticket
+  SELECT * INTO ticket FROM tbl_ticket where id = param_ticket_id AND (array_length(param_ticket_types, 1) IS NULL OR (array_position(param_ticket_types, ticket_type) IS NOT NULL));
```
Comment (j-licht) [inline src/Application/Migrations/15_function_ticket_state.sql]: I'm not sure, if it useful or not to return at this point.
That means also if children or depending tickets matches the filter, they will not be reset if the parent ticket doesn't match.

Comment (pegro) [reply src/Application/Migrations/15_function_ticket_state.sql]: I mean, missing an actual use-case I don't have a strong opinion on this.
I could see both cases (returning or not) being useful.
But not being exposed by an API it is even less useful ;)

Comment (j-licht) [reply src/Application/Migrations/15_function_ticket_state.sql]: So add a parameter to the function to control the behavior?

Comment (pegro) [reply src/Application/Migrations/15_function_ticket_state.sql]: Mh, no. Leave it as is, add the parameters to the RPC API and model and let us merge this PR.
When we feel we need to control this behavior further, we can change the code again, no?

**Comment by j-licht** [comment]
next try

File: src/Application/Model/Ticket.php [hunk 35]
```diff
@@ -851,6 +851,36 @@ public function resetSource(Comment $comment = null) {
 			
 			return true;
 		}
+
+		public function reset($handle_id = null, $ticket_types = []) {
+			$tickets = Database::$Instance->query(
+				'SELECT * FROM ticket_reset(?,?)',
+				[
+					$this['id'],
+					'{' . implode(',', $ticket_types) . '}'
```
Comment (pegro) [inline src/Application/Model/Ticket.php]: @jjeising Does this work properly? Enough escaping etc. so this won't lead to SQL injections?

Comment (jjeising) [reply src/Application/Model/Ticket.php]: Sorry, I've missed this one. This uses [PDO::prepare](http://php.net/manual/en/pdo.prepare.php), so if PostgreSQL supports arrays in prepared parameters (as it was hopefully tested by @j-licht ) this should be fine. Would be interesting to test, what happens if a user supplied ticket_type contains braces.

Comment (j-licht) [reply src/Application/Model/Ticket.php]: When I add a brace or comma, I got:
`PHP Fatal error:  Uncaught Exception: ERROR:  malformed array literal`

Comment (jjeising) [reply src/Application/Model/Ticket.php]: Ok. There are probably some other cases an user could trigger a SQL Exception (numeric values or similar), but in this case I would feel better if all ticket types fed into this query exist.

Comment (j-licht) [reply src/Application/Model/Ticket.php]: @pegro are you fine with this SQL building?

File: src/Application/Model/Ticket.php [hunk 36]
```diff
@@ -851,6 +851,36 @@ public function resetSource(Comment $comment = null) {
...
+				if ($ticket['follows_meta_ticket']) {
+					$entry['comment'] = 'Reset due to meta ticket reset.';
+				}
+				if ($ticket['follows_encoding_ticket']) {
+					$entry['comment'] = 'Reset due to ticket reset of depending encoding profile';
+				}
+
+				LogEntry::create($entry);
+			}
+			return count($tickets);
```
Comment (pegro) [inline src/Application/Model/Ticket.php]: If you push another version anyway, I'd like to have that count number also returned by the API.

File: src/Application/Model/Ticket.php [hunk 37]
```diff
@@ -851,6 +851,42 @@ public function resetSource(Comment $comment = null) {
 			
 			return true;
 		}
+
+		public function reset($handle_id = null, $ticket_types = []) {
+			foreach ($ticket_types as $ticket_type) {
+				if (!ctype_alnum($ticket_type)) {
+					throw new ModelException("ticket_types can only contain alpha numeric characters.");
```
Comment (j-licht) [inline src/Application/Model/Ticket.php]: I'm not sure whether this is the correct Exception.

Comment (jjeising) [reply src/Application/Model/Ticket.php]: This should be fine.

**Comment by jjeising** [comment]
We should probably add a database migration file for the new functions?

**Comment by a-tze** [comment]
absolutely, at least it was always done this way [tm]
